src/adc.rs

```diff
@@ -0,0 +1,636 @@
+                        resolution: Resolution::SIXTEENBIT,
+                        lshift: AdcLshift::default(),
+                    };
+                    s.enable_clock(&mut ccdr.$ahb, &mut ccdr.d3ccipr);
+                    s.power_down();
+                    s.reset(&mut ccdr.$ahb);
+                    s.power_up(delay);
+                    s.disable();
+                    s.preconfigure();
+                    s.calibrate();
+                    s.enable();
+                    s.configure();
+                    s
+                }
+
+                /// Save current ADC config
+                pub fn save_cfg(&mut self) -> StoredConfig {
+                    StoredConfig(self.get_sample_time(), self.get_resolution(), self.get_lshift())
+                }
+
+                /// Restore saved ADC config
+                pub fn restore_cfg(&mut self, cfg: StoredConfig) {
+                    self.set_sample_time(cfg.0);
+                    self.set_resolution(cfg.1);
+                    self.set_lshift(cfg.2);
+                }
+
+                /// Reset the ADC config to default, return existing config
+                pub fn default_cfg(&mut self) -> StoredConfig {
+                    let cfg = self.save_cfg();
+                    self.set_sample_time(AdcSampleTime::default());
+                    self.set_resolution(Resolution::SIXTEENBIT);
+                    self.set_lshift(AdcLshift::default());
+                    cfg
+                }
+
+                /// Get ADC samping time
+                pub fn get_sample_time(&self) -> AdcSampleTime {
+                    self.sample_time
+                }
+
+                /// Get ADC sampling resolution
+                pub fn get_resolution(&self) -> Resolution {
+                    self.resolution
+                }
+
+                /// Get ADC lshift value
+                pub fn get_lshift(&self) -> AdcLshift {
+                    self.lshift
+                }
+
+                /// Set ADC sampling time
+                ///
+                /// Options can be found in [AdcSampleTime](crate::adc::AdcSampleTime).
+                pub fn set_sample_time(&mut self, t_samp: AdcSampleTime) {
+                    self.sample_time = t_samp;
+                }
+
+                /// Set ADC sampling resolution
+                pub fn set_resolution(&mut self, res: Resolution) {
+                    self.resolution = res;
+                }
+
+                /// Set ADC lshift
+                ///
+                /// LSHIFT[3:0] must be in range of 0..=15
+                pub fn set_lshift(&mut self, lshift: AdcLshift) {
+                    self.lshift = lshift;
+                }
+
+                /// Returns the largest possible sample value for the current settings
+                pub fn max_sample(&self) -> u32 {
+                    ((1 << self.get_resolution().bits() as u32) - 1) << self.get_lshift().value() as u32
+                }
```

Comment:
This is broken (currently returns 0 for 16-bit). Try implementing a `number_of_bits` method to use here.

```rust
trait NumberOfBits {
    fn number_of_bits(&self) -> u32;
}
impl NumberOfBits for Resolution {
    fn number_of_bits(&self) -> u32 {
        match *self {
            Resolution::EIGHTBIT => 8,
            Resolution::TENBIT => 10,
            Resolution::TWELVEBIT => 12,
            Resolution::FOURTEENBIT => 14,
            _ => 16,
        }
    }
}
```